### DISCUSSION TITLE
Information on Linux drivers for PolarFire SoC peripherals

### DISCUSSION THREAD

**Discussion by ManishaI31214** [polarfire-soc/polarfire-soc-documentation#171] [Linux drivers for PolarFire Soc] [answered]
Can you kindly provide a table corresponding to the PolarFire SoC peripherals and the specific Linux drivers supporting them?
For example:  I2C IP <-> driver location in linux kernel source tree

**Comment by ConchuOD** [comment, marked as answer]
Hey Manisha,
Hopefully I am interpreting your question correctly, you want to figure out which drivers are used by specific peripherals?

On RISC-V the Linux kernel uses Device Tree to figure out what drivers it needs to use, based on a file we provide at boot time.
This file can be found in `arch/riscv/boot/dts/microchip`. The one for icicle kit is called `mpfs-icicle-kit.dts`. In this file there are a load of "nodes", each containing a "compatible" string (or series of strings!). This string defines what drivers are allowed to bind against a devicetree entry & goes in order of precedence from left to right. It is possible that drivers will not exist for the left-most entry, in which case a less specific driver will be used.

A nice example of this is if we look in `arch/riscv/boot/dts/microchip` at the mmc:
```
		mmc: mmc@20008000 {
			compatible = "microchip,mpfs-sd4hc", "cdns,sd4hc";
			reg = <0x0 0x20008000 0x0 0x1000>;
			interrupt-parent = <&plic>;
			interrupts = <88>;
			clocks = <&clkcfg CLK_MMC>;
			max-frequency = <200000000>;
			status = "disabled";
		};
```
Looking in the drivers subdirectory with [ripgrep](https://github.com/BurntSushi/ripgrep) returns nothing:
```
rg "microchip,mpfs-sd4hc" drivers/
```
because there is no specific driver for our mmc controller. Then using the fallback compatible:
```
rg "cdns,sd4hc" drivers/
drivers/mmc/host/sdhci-cadence.c
458:    { .compatible = "cdns,sd4hc" },
```
We can see that this string is supported by a driver, in this case the generic driver for Cadence IP.
It's also possible that multiple drivers may match against a compatible string, in which case build order may decide which is loaded!
For i2c it is a similar story:
```
rg "microchip,mpfs-i2c" drivers/
drivers/i2c/busses/i2c-microchip-corei2c.c
460:    { .compatible = "microchip,mpfs-i2c" },
```

**Reply by ConchuOD** [reply]
Hey @ManishaI31214 did this help at all?

**Comment by ManishaI31214** [comment]
Hi Conor,

Yes, this did help the customer and no more questions from customer.
